From: Tim 'mithro' Ansell <me@mith.ro>
Subject: Revert "Fold smarter _choose_fallback_pubkey_algorithm into base AuthHandler"

Restore legacy SSH support removed upstream, for talking to old
devices that cannot be upgraded. Shipped only in the separate
paramiko_legacy module; python3-paramiko is untouched.

Origin: vendor, reverts https://github.com/paramiko/paramiko/commit/d99939c27ad8e35ec278d11581d155c309757182
Forwarded: not-needed
---
diff --git a/paramiko/auth_handler.py b/paramiko/auth_handler.py
index b927a2a2..f088912f 100644
--- a/paramiko/auth_handler.py
+++ b/paramiko/auth_handler.py
@@ -294,18 +294,15 @@ class AuthHandler:
         return self.transport._key_info[algorithm](Message(keyblob))
 
     def _choose_fallback_pubkey_algorithm(self, key_type, my_algos):
-        msg = "Server did not send a server-sig-algs list; defaulting to something in our preferred algorithms list"  # noqa
-        self._log(DEBUG, msg)
-        noncert_key_type = key_type.replace("-cert-v01@openssh.com", "")
-        if key_type in my_algos or noncert_key_type in my_algos:
-            actual = key_type if key_type in my_algos else noncert_key_type
-            msg = f"Current key type, {actual!r}, is in our preferred list; using that"  # noqa
-            algo = actual
-        else:
-            algo = my_algos[0]
-            msg = f"{key_type!r} not in our list - trying first list item instead, {algo!r}"  # noqa
-        self._log(DEBUG, msg)
-        return algo
+        # Fallback: first one in our (possibly tweaked by caller) list
+        pubkey_algo = my_algos[0]
+        msg = "Server did not send a server-sig-algs list; defaulting to our first preferred algo ({!r})"  # noqa
+        self._log(DEBUG, msg.format(pubkey_algo))
+        self._log(
+            DEBUG,
+            "NOTE: you may use the 'disabled_algorithms' SSHClient/Transport init kwarg to disable that or other algorithms if your server does not support them!",  # noqa
+        )
+        return pubkey_algo
 
     def _finalize_pubkey_algorithm(self, key_type):
         # Short-circuit for non-RSA keys
@@ -1084,3 +1081,18 @@ class AuthOnlyHandler(AuthHandler):
             m.add_string(submethods)
 
         return self.send_auth_request(username, "keyboard-interactive", finish)
+
+    # NOTE: not strictly 'auth only' related, but allows users to opt-in.
+    def _choose_fallback_pubkey_algorithm(self, key_type, my_algos):
+        msg = "Server did not send a server-sig-algs list; defaulting to something in our preferred algorithms list"  # noqa
+        self._log(DEBUG, msg)
+        noncert_key_type = key_type.replace("-cert-v01@openssh.com", "")
+        if key_type in my_algos or noncert_key_type in my_algos:
+            actual = key_type if key_type in my_algos else noncert_key_type
+            msg = f"Current key type, {actual!r}, is in our preferred list; using that"  # noqa
+            algo = actual
+        else:
+            algo = my_algos[0]
+            msg = f"{key_type!r} not in our list - trying first list item instead, {algo!r}"  # noqa
+        self._log(DEBUG, msg)
+        return algo
diff --git a/tests/auth.py b/tests/auth.py
index 70335386..14a57f1f 100644
--- a/tests/auth.py
+++ b/tests/auth.py
@@ -25,6 +25,7 @@ from paramiko import (
     PKey,
     PrivateKey,
     RSAKey,
+    ServiceRequestingTransport,
     SourceResult,
     SSHException,
 )
@@ -149,6 +150,79 @@ class AuthHandler_:
             assert "Authentication timeout" in str(info.value)
 
 
+class AuthOnlyHandler_:
+    def _server(self, *args, **kwargs):
+        kwargs.setdefault("transport_factory", ServiceRequestingTransport)
+        return server(*args, **kwargs)
+
+    class fallback_pubkey_algorithm:
+        def key_type_algo_selected_when_no_server_sig_algs(self):
+            privkey = RSAKey.from_private_key_file(_support("rsa.key"))
+            # Server pretending to be an apparently common setup:
+            # - doesn't support (or have enabled) sha2
+            # - also doesn't support (or have enabled) server-sig-algs/ext-info
+            # This is the scenario in which Paramiko has to guess-the-algo, and
+            # where servers that don't support sha2 or server-sig-algs can give
+            # us trouble.
+            # TODO: ok maybe I need to reinstate _disable_sha2_pubkey, but it
+            # _might_ make sense to choose some other key type to disable, if
+            # it's more sensible now - surely the modern version of this
+            # scenario would be disabling like, ecdsa or something?
+            server_init = dict(_disable_sha2_pubkey, server_sig_algs=False)
+            with self._server(
+                pubkeys=[privkey],
+                connect=dict(pkey=privkey),
+                server_init=server_init,
+                catch_error=True,
+            ) as (tc, ts, err):
+                # Auth did work
+                assert tc.is_authenticated()
+                # Selected ssh-rsa, instead of first-in-the-list (rsa-sha2-512)
+                # TODO: this needs to be selecting some other fallback now,
+                # presumably whatever is now first in one of our attribute
+                # lists
+                assert tc._agreed_pubkey_algorithm == "ssh-rsa"
+
+        def key_type_algo_selection_is_cert_suffix_aware(self):
+            # This key has a cert next to it, which should trigger cert-aware
+            # loading within key classes.
+            privkey = PKey.from_path(_support("rsa.key"))
+            server_init = dict(_disable_sha2_pubkey, server_sig_algs=False)
+            with self._server(
+                pubkeys=[privkey],
+                connect=dict(pkey=privkey),
+                server_init=server_init,
+                catch_error=True,
+            ) as (tc, ts, err):
+                assert not err
+                # Auth did work
+                assert tc.is_authenticated()
+                # Selected expected cert type
+                assert (
+                    tc._agreed_pubkey_algorithm
+                    # TODO: this needs to be choosing a sha2 or w/e now
+                    == "ssh-rsa-cert-v01@openssh.com"
+                )
+
+        def uses_first_preferred_algo_if_key_type_not_in_list(self):
+            # This is functionally the same as legacy AuthHandler, just
+            # arriving at the same place in a different manner.
+            privkey = RSAKey.from_private_key_file(_support("rsa.key"))
+            server_init = dict(_disable_sha2_pubkey, server_sig_algs=False)
+            with self._server(
+                pubkeys=[privkey],
+                connect=dict(pkey=privkey),
+                server_init=server_init,
+                # TODO: this is outdated now, disable something else, probably
+                # whatever is /now/ first in the list instead of ssh-rsa
+                client_init=_disable_sha1_pubkey,  # no ssh-rsa
+                catch_error=True,
+            ) as (tc, ts, err):
+                assert not tc.is_authenticated()
+                assert isinstance(err, AuthenticationException)
+                assert tc._agreed_pubkey_algorithm == "rsa-sha2-512"
+
+
 class SHA2SignaturePubkeys:
     def pubkey_auth_honors_disabled_algorithms(self):
         privkey = RSAKey.from_private_key_file(_support("rsa.key"))
@@ -202,6 +276,7 @@ class SHA2SignaturePubkeys:
         ) as (tc, _):
             assert tc.is_authenticated()
 
+    # TODO: isn't this duplicating some of the earlier tests? if not, update it
     def first_client_preferred_algo_used_when_no_server_sig_algs(self):
         privkey = RSAKey.from_private_key_file(_support("rsa.key"))
         # Server pretending to be an apparently common setup:
@@ -210,10 +285,6 @@ class SHA2SignaturePubkeys:
         # This is the scenario in which Paramiko has to guess-the-algo, and
         # where servers that don't support sha2 or server-sig-algs give us
         # trouble.
-        # TODO: disabling sha2 _in this case_ might be ok, _as long as_ we
-        # update to account for the client's new first preference being
-        # something besides ssh-rsa (and where the client also has that
-        # disabled?)
         server_init = dict(_disable_sha2_pubkey, server_sig_algs=False)
         with server(
             pubkeys=[privkey],
